onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlSTFT.h

```diff
@@ -0,0 +1,610 @@
+#pragma once
+
+#include "DmlDFT.h"
+
+enum DmlSTFTKernelInputIndex : uint32_t
+{
+    Signal = 0,
+    FrameStep,
+    Window,
+    FrameLength
+};
+
+// Helper to derive dimensions and attributes from either the STFT shape inferrer or the STFT kernel constructor.
+//
+// Type             | Shape Inference                         | Kernel Constructor
+// -----------------|-----------------------------------------|-----------------------------------------
+// ContextInterface | IMLOperatorShapeInferenceContextPrivate | IMLOperatorKernelCreationContextPrivate
+// ShapeInterface   | IMLOperatorShapeInferenceContextPrivate | IMLOperatorTensorShapeDescription
+template <typename ContextInterface, typename ShapeInterface>
```

Comment:
>typename ContextInterface, typename ShapeInterface [](http://example.com/codeflow?start=10&length=50)

I got tired duplicating nearly identical functions and having multiple copies of them in the code (yay template bloat) and so in OperatorAuthorHelper\OperatorHelper.h started passing `IShapeInformationAdapter` and `IKernelInformationAdapter` rather than the templated type. Would that work here?

We should talk with Jeff about unifying these somehow with a common interface, because they duplicate so many common methods.

Reply:
Ah, didn't notice those. It took some refactoring to get these to work, but it's a bit cleaner.

Reply:
Thanks.